dvc/data/stage.py

```diff
@@ -247,27 +250,106 @@ def _get_staging(odb: "ObjectDB") -> "ObjectDB":
     return ReferenceObjectDB(fs, fs_path, state=state)
 
 
-def _load_from_state(odb, staging, fs_path, fs, name):
+def _try_load(
+    odb: "ObjectDB",
+    hash_info: "HashInfo",
+) -> Optional["HashFile"]:
     from dvc.objects.errors import ObjectFormatError
 
     from . import check, load
+
+    try:
+        obj = load(odb, hash_info)
+        check(odb, obj, check_hash=False)
+    except (ObjectFormatError, FileNotFoundError):
+        obj = None
+
+    return obj
+
+
+def _load_raw_dir_obj(odb: "ObjectDB", hash_info: "HashInfo") -> "Tree":
+    from dvc.objects.errors import ObjectFormatError
+
+    from .tree import Tree
+
+    try:
+        tree = Tree.load(odb, hash_info.as_raw())
+        tree.check(odb)
+    except ObjectFormatError:
+        raise FileNotFoundError
```

Comment:
```suggestion
        raise FileNotFoundError from exc
```

Reply:
While we are here, 
```suggestion
        raise FileNotFoundError(errno.ENOENT, "No such object", odb.hash_to_path(hash_info.value))
```

Or, just use `hash_info.value` for simplicity.